File: 2022/3-longest-substring-without-repeating.cpp
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

// Better solution is to use sliding window!
#if 1
class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        if (s.empty()) { return 0; }
        int longest = 1;
        // O(n^2)
        for (int i = 0; i < s.length(); i++)
        {
            std::unordered_set<char> substring;
            for (int j = i; j < s.length(); j++)
            {
                if (substring.find(s[j]) == substring.end())
                {
                    substring.insert(s[j]);
                }
                else
                {
                    break;
                }
            }
            if (longest < substring.size())
            {
                longest = substring.size();
            }
        }
        return longest;
    }
};
#else
class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {

    }
};
#endif
int main()
{
    Solution s;
    std::vector<std::string> tests = {"au", "abcabcbb", "bbbbb", "pwwkew"};
    for (std::string& string : tests)
    {
        std::cout << string << ": " << s.lengthOfLongestSubstring(string) << std::endl;
    }
}

File: top150/python/3_longest_substring_without_repeating.py
from typing import List

# https://leetcode.com/problems/longest-substring-without-repeating-characters/
# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         # Using a dictionary to track whether there are duplicates in the range
#         dict = {}
#         max_len = 0
#         left = 0
#         for r in range(len(s)):

#             if s[r] in dict:
#                 dict[s[r]] += 1
#             else:
#                 dict[s[r]] = 0

#             while len(dict) != (r - left + 1): # While the range contains duplicates
#                 if dict[s[left]] == 0:
#                     del dict[s[left]]
#                 else:
#                     dict[s[left]] -= 1
#                 left += 1

#             max_len = max(max_len, r - left + 1)
#         return max_len
    

# Previous solution works, but optimise!
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        # Using a dictionary to track the last index we saw a given character at
        seen = {}
        max_len = 0
        left = 0
        for r in range(len(s)):
            char = s[r]

            # Check the last index we saw this character at and if it is within the window
            if char in seen and seen[char] >= left:
                # If it is, we can jump left pointer to the last index of the duplicate character + 1
                # Because any window with left pointer = (r) up to (last index of duplicate) contains a duplicate!
                left = seen[char] + 1
            else:
                max_len = max(max_len, r - left + 1)

            seen[char] = r

        return max_len

import unittest

class Test(unittest.TestCase):

    def test_basic(self):
        s = Solution()
        self.assertEqual(s.lengthOfLongestSubstring("abcabcbb"), 3)
        self.assertEqual(s.lengthOfLongestSubstring("bbbbb"), 1)
        self.assertEqual(s.lengthOfLongestSubstring("pwwkew"), 3)

if __name__ == "__main__":
    unittest.main()
